feat: 구매자는 진행 중인 경매를 입찰할 수 있다.(뼈대코드만)

minseok-oh opened the pull request:
## 📄 Summary

- `Payment` 클래스를 구현하여 입찰 서비스 로직을 구현했습니다.
- `Payment` 도메인은 루트 도메인으로 입찰의 시작점을 담당하며 입찰 과정 중에 다른 도메인들을 사용합니다.
- `Member` 도메인에 `Point` 도메인을 필드로 넣어 구현했습니다.

## 🙋🏻 More

- 7조 화이팅~ 주말에도 열심히~!

close #24

minseok-oh committed:
refactor: payment 계층 생성에 따라 BidHistory 위치를 변경합니다

minseok-oh committed:
feat: 페이먼츠 도메인 클래스 생성

minseok-oh committed:
feat: 포인트 도메인 클래스 생성

minseok-oh committed:
feat: 멤버 필드에 포인트 추가

- 멤버 없이 포인트는 사용되지 않기 때문에 필드에 추가합니다.

minseok-oh committed:
feat: 가격과 재고를 가지도록 auction info dto 재정의

- 입찰 진행 중에 경매 정보가 필요하여 정의합니다.

minseok-oh committed:
feat: 구매자, 판매자 클래스 생성자에 포인트 추가

- 멤버 클래스의 포인트 필드 추가에 따라 하위 클래스들에 포인트를 추가합니다.

minseok-oh committed:
feat: 멤버 팩토리 메서드에 포인트 추가

- 멤버 클래스의 포인트 필드 추가에 따라 팩토리 메서드에 포인트를 추가합니다.

minseok-oh committed:
feat: 페이먼츠 입찰 서비스 로직 구현

- canBuy 메서드는 Buyer 클래스에 구현되어 dto가 아닌 멤버 도메인을 인자로 받았습니다.
- AuctionService에 의존을 하고 있어 테스트 코드 틀만 구현했습니다.

minseok-oh commented on src/main/java/com/wootecam/luckyvickyauction/core/payment/service/PaymentService.java:
```diff
@@ -0,0 +1,30 @@
+public class PaymentService {
+    private final AuctionService auctionService;
+
+    public void submitBid(Member signInMember, long auctionId, int quantity) {
+        if (!signInMember.isBuyer()) {
+            throw new BadRequestException("구매자만 입찰을 할 수 있습니다.", ErrorCode.P000);
+        }
+
+        AuctionInfo auctionInfo = auctionService.getAuction(auctionId);
+        if (auctionInfo.quantity() < quantity) {
+            throw new BadRequestException("남아있는 재고보다 더 많은 입찰은 불가능합니다.", ErrorCode.P001);
+        }
+
+        Buyer signInBuyer = (Buyer) signInMember;
+        if (signInBuyer.canBuy(auctionInfo.price(), quantity)) {
+            throw new BadRequestException("포인트가 부족합니다.", ErrorCode.P002);
+        }
+    }
+}
```
- 로그인한 유저가 세션에 들어있기 때문에 dto로 변환하지 않았습니다
- JPA를 사용할 때 영속성 컨텍스트 문제가 생기지 않을 것으로 예상됩니다
- 다른 의견이 있다면 부탁드립니다!

yudonggeun replied:
Member 클래스는 jpa entity로 사용할 계획인가요?

HiiWee replied:
`@MappedSuperClass`로 두고 사용하지 않을까 생각됩니다!

HiiWee commented on src/main/java/com/wootecam/luckyvickyauction/core/auction/dto/AuctionInfo.java:
```diff
@@ -1,4 +1,8 @@
 package com.wootecam.luckyvickyauction.core.auction.dto;
 
-public record AuctionInfo() {
+/**
+ * @param price    현재 단일 상품의 가격
+ * @param quantity 현재 경매에 구매할 수 있는 재고 개수
+ */
+public record AuctionInfo(Long price, Long quantity) {
```
이 부분은 일단 임의로 정의했습니당 네이밍이나, 잘못된 필드가 기입되어 있다면 알려주세요!

HiiWee replied:
wrapper는 의도한건 아닌데,, 자동 생성될때 들어갔네요

yudonggeun replied:
Long -> long 으로 기본형을 쓰면 좋을 것 같아요!

HiiWee replied:
반영하겠습니다~

chhs2131 commented on src/main/java/com/wootecam/luckyvickyauction/core/member/domain/Member.java:
```diff
@@ -2,19 +2,26 @@
+    private Long id;
     private String signInId;
     private Role role;
+    protected Point point;
```
혹시 Point 변수의 접근제어자는 왜 protected 일까요? Buyer가 Seller의 포인트를 훔쳐가는 그런느낌인가요?! 🔥

HiiWee replied:
`Buyer`와 `Seller`가 포인트를 쓰는 방식이 각각 다르기 때문에 해당 클래스에서 처리할 수 있게 protected로 정의했어요!

yudonggeun commented on src/test/java/com/wootecam/luckyvickyauction/core/payment/service/PaymentServiceTest.java:
```diff
@@ -0,0 +1,56 @@
+import org.junit.jupiter.api.Test;
+
+// TODO: AuctionService 세부 사항 결정되면 테스트
+class PaymentServiceTest {
```
테스트 정의 좋았다 👍

HiiWee reviewed (changes requested):
테스트 request changes 요청

HiiWee committed:
refactor: wrapper class to primitive type

- 수량, 가격은 null이 들어가지 않음

HiiWee committed:
test

HiiWee committed:
test: branch rules test

HiiWee committed:
feat: 입찰 검증을 마치고 실제 입찰을 완료한다.

- 아직 BidHistory 도메인 설계가 완료되지 않아 흐름을 담은 로직만 존재
- 구매 수량 * 가격에 대한 검증 로직을 Point 내부로 변경

HiiWee commented on src/main/java/com/wootecam/luckyvickyauction/core/payment/service/PaymentService.java:
```diff
@@ -0,0 +1,38 @@
+        if (auctionInfo.quantity() < quantity) {
+            throw new BadRequestException("남아있는 재고보다 더 많은 입찰은 불가능합니다.", ErrorCode.P001);
+        }
+
+        // TODO 입찰 내역 도메인 설계 필요 -> id값을 가질건지, 도메인을 가질건지
+        Buyer signInBuyer = (Buyer) signInMember;
+        signInBuyer.usePoint(auctionInfo.price(), quantity);
+
+        BidHistory bidHistory = new BidHistory();
+        bidHistoryRepository.save(bidHistory);
+        memberRepository.save(signInMember);
+    }
+}
```
@yudonggeun @chhs2131 
새로 추가된 로직부분인데 아직 BidHistory 도메인을 정의하지 않아서 껍데기만 남겨두었어요 확인부탁드립니당

chhs2131 replied:
확인했습니다!

chhs2131 reviewed (approved):
주말에도 고생이 많습니다~
